README
NAME
    Encode::Escape - Perl extension for Encodings of various escape
    sequences

SYNOPSIS
      use Encode::Escape;

      $escaped_ascii = "Perl\\tPathologically Eclectic Rubbish Lister\\n";
      $ascii = decode 'ascii-escape', $escaped_ascii;
  
      # Now, $ascii is equivalent to 
      # double quote string "Perl\tPathologically Eclectic Rubbish Lister\n"

      $escaped_unicode = "Perl \\x{041F}\\x{0435}\\x{0440}\\x{043B} \\x{D384}" 
      $string = decode 'unicode-escape', $escaped_unicode;

      # Now, $string is equvialent to 
      # double quote string "Perl \x{041F}\x{0435}\x{0440}\x{043B} \x{D384}" 

    It may looks non-sense. Here's another case.

    If you have a text data file 'ascii-escape.txt'. It contains a line:

      Perl\tPathologically Eclectic Rubbish Lister\n

    And you want to use it as if it were a normal double quote string in
    source code. Try this:

      open(FILE, 'ascii-escape.txt');
      while(<FILE>) {
        chomp;
        print decode 'ascii-escape', $_;
      }

DESCRIPTION
    Encode::Escape module is a wrapper class for encodings of escape
    sequences.

    It is NOT for an escape-based encoding (eg. ISO-2022-JP). It is for
    encoding/decoding escape sequences, generally used in source codes.

    Many programming languages, markup languages, and typesetting languages
    provide methods for encoding special (or functional) characters and
    non-keyboard symbols in the form of escape sequence. That's what I
    concern.

    For example, Encode::Escape::Unicode converts (encodes) Unicode
    characters into escape sequences representing it: "\x{AC00}" to the
    first Hangul syllable Ka, "\x{042F}" to Cyrillic capital letter Ya, etc.

    Yes, you're right. There already exist many modules. See String::Escape,
    Unicode::Escape, TeX::Encode, HTML::Mason::Escape,
    Template::Plugin::XML::Escape, URI::Escape, etc.

    But for some reason I need to do it in a different way. There is more
    than one way to do it! After that, I asked myself if this module is
    useful. May be not except for me. At this time, Zhuangzi reminds me that
    "the useless has its use".

  Escape Sequences
   Character Escape Codes
    ASCII defines 128 characters: 33 non-printing control characters (0x00
    -- 0x1f, 0x7f) and 95 printable characters (0x20 -- 0x7e).

    Character Escape Codes in C programming language provide a method to
    express control characters, using only printable ones. These are
    accepted by Perl and many other languages.

     CEC    HEX     Description
     ---    ----    --------------
     \0     00      Null character
     \a     07      Bell
     \b     08      Backspace
     \t     09      Horizontal Tab
     \n     0a      Line feed
     \v     0b      Vertical Tab
     \f     0c      Form feed
     \r     0d      Carriage return

    Programming languages provide escape sequences for printable characters,
    which have significant meaning in that language. Otherwise, it would be
    harder to print them literally.

     ESC    HEX     Description
     ---    ----    ---------------
     \"     22      double quote 
     \\     52      backslash

    Refer to ASCII, Escape character, Escape sequence at
    <http://en.wikipedia.org>, for more details.

   Perl Escape Sequences
    Perl use backslash as an escape character.

    These work in normal strings and regular expressions except \b.

     ESC       Description
     ---       --------------------------
     \a        Alarm (beep)
     \b        Backspace
     \e        Escape
     \f        Formfeed
     \n        Newline
     \r        Carriage return
     \t        Tab
     \037      Any octal ASCII value
     \x7f      Any hexadecimal ASCII value
     \x{263a}  A wide hexadecimal value
     \cx       Control-x
     \N{name}  name is a name for the Unicode character (use charnames)
 
    The following escape sequences are available in constructs that
    interpolate.

     \l        Lowercase next character
     \u        Titlecase next character
     \L        Lowercase until \E
     \U        Uppercase until \E
     \E        End case modification
 
    In regular expresssions:

     \b        An assetion, not backspace, except in a character class
     \Q        Disable pattern metacharacters until \E

    Unlike C and other languages, Perl has no \v escape sequence for the
    vertical tab (VT - ASCII 11).

    For constructs that do interpolate, variable begining with "$" or "@"
    are interpolated.

     \$        Dollar Sign
     \@        Ampersand
     \"        Print double quotes
     \         Escape next character if know otherwise print

    See perlerref, perlop

   Python Escape Sequences
     \newline   Ignored
     \\         Backslash (\)
     \'         Single quote (')
     \"         Double quote (")
     \a         ASCII Bell (BEL)
     \b         ASCII Backspace (BA)
     \f         ASCII Formfeed (FF)
     \n         ASCII Linefeed (LF)
     \N{name}   Character named 'name' in the Unicode database
     \r         ASCII Carriage Return (CR)
     \t         ASCII Horizontal Tab (TAB)
     \uxxxx     Character with 16-bit hex value xxxx (Unicode only)
     \Uxxxxxxxx Character with 32-bit hex value xxxxxxxx (Unicode only)
     \v         ASCII Vertical Tab (VT)
     \ooo       Character with octal value ooo
     \xhh       Character with hex value hh

    See <http://docs.python.org/ref/strings.html>

    Unicode escape sequences in the form of "\u"*xxxx* are used in Java,
    Python, C#, JavaScript. Unicode::Escape module implements it.

   LaTeX Escapes
    TeX::Encode implements encodings of LaTeX escapes. It converts (encodes)
    utf8 string to LaTeX escapes.

SEE ALSO
The useless has its use
    Hui Tzu said to Chuang Tzu, "Your words are useless!"

    Chuang Tzu said, "A man has to understand the useless before you can
    talk to him about the useful. The earth is certainly vast and broad,
    though a man uses no more of it than the area he puts his feet on. If,
    however, you were to dig away all the earth from around his feet until
    you reached the Yellow Springs, then would the man still be able to make
    use of it?"

    "No, it would be useless," said Hui Tzu.

    "It is obvious, then," said Chuang Tzu, "that the useless has its use."

AUTHOR
    You Hyun Jo, <you at cpan dot org>

COPYRIGHT AND LICENSE
    Copyright (C) 2007 by You Hyun Jo

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.8 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Encode::Escape.

0.10 2007/12/05 06:20:08 you 
   - first upload to CPAN.

   Encode::Escape::ASCII
	- implemented basic functions.

   Encode::Escape::Unicode 
	- implemented basic functions
	- implemented mode system
		- python(or java) mode enable to use "\uxxxx" unicode escape seqs.

0.01  Sun Dec  2 21:45:59 2007
	- original version; created by h2xs 1.23 with options
		-X -n Encode::Escape



lib/Encode/Escape/ASCII.pm
# Encoding of ASCII Escape Sequences (or Escaped ASCII)

# $Id: ASCII.pm,v 1.17 2007-12-05 14:16:54+09 you Exp $

package Encode::Escape::ASCII;

our $VERSION  = do { q$Revision: 1.17 $ =~ /\d+\.(\d+)/; sprintf "%.2f", $1 / 100  };

use 5.008008;
use strict;
use warnings;

use Encode::Encoding;
use base qw(Encode::Encoding);

__PACKAGE__->Define(qw/ascii-escape ascii_escape/);

sub import {

    require Encode;
    Encode->export_to_level(1, @_);
}

sub enmode ($$) {
    my ($class, $mode) = @_;
    
}

sub demode ($$) {
    my ($class, $mode) = @_;
}


sub encode($$;$) {
    my ($obj, $str, $chk) = @_;
    my $escaped = escape($str);
    return $escaped;
}


sub decode($$;$) {
    my ($obj, $str, $chk) = @_;
    my $unescaped = unescape($str);
    return $unescaped;
}

my %ESCAPED = ( 
    "\\" => '\\', 
    "\r" => 'r', 
    "\n" => 'n', 
    "\t" => 't', 
    "\a" => 'a',
    "\b" => 'b',
    "\e" => 'e',
    "\f" => 'f',
    "\"" => '"',
    "\$" => '$',
    "\@" => '@',
);

my %UNESCAPED = ( reverse %ESCAPED );

sub chr2hex {
    my($c) = @_;
    if ( ord($c) < 128 ) {
        return sprintf("\\x%02x", ord($c));
    }
    else {
        require Carp;
        Carp::croak (
            "'ascii-escape' codec can't encode character: ordinal " . ord($c)
        );
    }
}


sub escape ($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );
    s/([\a\b\e\f\r\n\t\"\\\$\@])/\\$ESCAPED{$1}/sg;
    s/([\x00-\x1f\x7f-\xff])/chr2hex($1)/gse;
    return $_;
}

sub hex2chr { 
    my($hex) = @_;
    if ( hex($hex) >= 0 and hex($hex) < 128) {
        return chr(hex($hex));
    }
    else {
        require Carp;
        Carp::croak(
            "'ascii-escape' codec can't decode escape sequence: " 
            . "\\x$hex (ordinal " . hex($hex) . ")"
        );
    }
}
sub oct2chr {
    my($oct) = @_;
    if ( oct($oct) >= 0 and oct($oct) < 128 ) {
        return chr(oct($oct));
    }
    else {
        require Carp;
        Carp::croak (
            "'ascii-escape' codec can't decode escape sequence: " 
            . "\\$oct (ordinal " . oct($oct). ")"
        );
    }
}

# $original_string = unprintable( $special_characters_escaped );
sub unescape ($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );

    s/((?:\A|\G|[^\\]))\\x([\da-fA-F]{1,2})/$1.hex2chr($2)/gse;
    s/((?:\A|\G|[^\\]))\\x\{([\da-fA-F]{1,4})\}/$1.hex2chr($2)/gse;
    s/((?:\A|\G|[^\\]))\\([0-7]{1,3})/$1.oct2chr($2)/gse;

    s/((?:\A|\G|[^\\]))\\([^aAbBeEfFrRnNtT\\\"\$\@])/$1$2/g;
    s/((?:\A|\G|[^\\]))\\([aAbBeEfFrRnNtT\\\"\$\@])/$1.$UNESCAPED{lc($2)}/gse;

    return $_;
}


1;
__END__
=head1 NAME

Encode::Escape::ASCII - Perl extension for Encoding of ASCII Escape Sequnces
(or Escaped ASCII)

=head1 SYNOPSIS

  use Encode::Escape::ASCII;

  $escaped = "Perl\\tPathologically Eclectic Rubbish Lister\\n";
  $string = decode 'ascii-escape', $escaped;

  # Now, $string is equivalent to "Perl\tPathologically Eclectic Rubish Lister\n";

  $escaped = "\\x65\\x50\\x6c\\x72\\x50\\x09\\x74\\x61\\x6f\\x68\\x6f\\x6c" .
             "\\x69\\x67\\x61\\x63\\x6c\\x6c\\x20\\x79\\x63\\x45\\x65\\x6c" .
             "\\x74\\x63\\x63\\x69\\x52\\x20\\x62\\x75\\x73\\x69\\x20\\x68" .
             "\\x69\\x4c\\x74\\x73\\x72\\x65\\x0a";
  $string = decode 'ascii-escape', $escaped;

  # Now, $string is equivalent to "Perl\tPathologically Eclectic Rubish Lister\n";

If you have a text data file 'ascii-escape.txt'. It contains a line:

  Perl\tPathologically Eclectic Rubbish Lister\n

And you want to use it as if it were a normal double quote string in source 
code. Try this:

  open(FILE, 'ascii-escape.txt');
  while(<FILE>) {
    chomp;
    print decode 'ascii-escape', $_;
  }

=head1 DESCRIPTION

L<Encode::Escape::ASCII> module implements escape-based encodings of ASCII.

Simply saying, it converts (decodes) escape sequences into ASCII chracters 
(0x00 -- 0x7f) and converts (encodes) non-printable control characters 
into escape sequences.

It supports only ASCII character. ASCII is called as low ASCII or 7-bit ASCII 
when one distinguishes it from various extended ASCII codes, i.e. high ASCII 
or 8-bit ASCII.

=head2 Supproted Escape Sequences

 Escape Sequcnes      Description
 ---------------      --------------------------
 \a                   Alarm (beep)
 \b                   Backspace
 \e                   Escape
 \f                   Formfeed
 \n                   Newline
 \r                   Carriage return
 \t                   Tab
 \000     - \177      octal ASCII value. \0, \00, and \000 are equivalent.
 \x00     - \x7f      hexadecimal ASCII value. \x0 and \x00 are equivalent.
 \x{0000} - \x{007f}  hexadecimal ASCII value. \x{0}, \x{00}, x\{000}, \x{0000}


 \\                   Backslash
 \$                   Dollar Sign
 \@                   Ampersand
 \"                   Print double quotes
 \                    Escape next character if known otherwise print

=head1 AUTHOR

you, E<lt>you at cpan dot orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by you 

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
# vi: set ts=4 sts=4 sw=4 et


lib/Encode/Escape/Unicode.pm
# Encoding of Unicode Escape Sequences (or Escaped Unicode)

# $Id: Unicode.pm,v 1.11 2007-12-05 14:16:38+09 you Exp $

package Encode::Escape::Unicode;

our $VERSION  = do { q$Revision: 1.11 $ =~ /\d+\.(\d+)/; sprintf "%.2f", $1 / 100  };

use 5.008008;
use strict;
use warnings;

use Encode::Encoding;
use base qw(Encode::Encoding);

__PACKAGE__->Define(qw/unicode-escape unicode_escape/);

sub import {

    __PACKAGE__->enmode('default');
    __PACKAGE__->demode('default');

    require Encode;
    Encode->export_to_level(1, @_);
}

our $enmode;
our $demode;
sub encoder($);
sub decoder($);

#
# == encoder/decoder modes ==
#
our %encoder = (
    undef   => \&perl_encoder,
    ''      => \&perl_encoder,
    default => \&perl_encoder,
    perl    => \&perl_encoder,
    java    => \&python_encoder,
    python  => \&python_encoder,
    csharp  => \&python_encoder,
);

our %decoder = (
    undef   => \&perl_decoder,
    ''      => \&perl_decoder,
    default => \&perl_decoder,
    perl    => \&perl_decoder,
    java    => \&python_decoder,
    python  => \&python_decoder,
    csharp  => \&python_decoder,
);

#
# == encode/decode ==
#

sub encode($$;$) {
    my ($obj, $str, $chk) = @_;
    $_[1] = '' if $chk;
    return encoder $str;
}


sub decode($$;$) {
    my ($obj, $str, $chk) = @_;
    $_[1] = '' if $chk;
    return decoder $str;
}

#
# == enmode/demode ==
#

sub enmode ($$) {
   my ($class, $mode) = @_;
   $mode = 'undef' unless defined $mode;
   unless (exists $encoder{$mode}) {
        require Carp;
        Carp::croak(
            "Unknown enmode '$mode' for encoding '" . $class->name() . "'"
        );
   }
   $enmode = $mode;
}

sub demode ($$) {
   my ($class, $mode) = @_;
   $mode = 'undef' unless defined $mode;
   unless (exists $decoder{$mode}) {
        require Carp;
        Carp::croak(
            "Unknown demode '$mode' for encoding '" . $class->name() . "'"
        );
   }
   $demode = $mode;
}


#
# = DATA AND SUBROUTINES FOR INTERNAL USE =
#


#
# == encoder/decoder ==
#

sub encoder($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );
    return $encoder{$enmode}->($_);
}
sub decoder($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );
    return $decoder{$demode}->($_);
}

#
# == enmode_encoder / demode_decoder ==
#

# default (perl) escape sequences
#
sub perl_encoder($) {

    local $_ = ( defined $_[0] ? $_[0] : '' );

    $_ = escape($_);
    s/([\x00-\x1f\x{7f}-\x{ffff}])/"\\x\{".uc(chr2hex($1))."\}"/gse;

    return $_;
}

sub perl_decoder($) {

    local $_ = ( defined $_[0] ? $_[0] : '' );

    s/((?:\A|\G|[^\\]))\\x([\da-fA-F]{1,2})/$1.hex2chr($2)/gse;
    s/((?:\A|\G|[^\\]))\\x\{([\da-fA-F]{1,4})\}/$1.hex2chr($2)/gse;

    return unescape($_);
}

# python (or java, c#) escape sequences 
#
sub python_encoder($) {

    local $_ = ( defined $_[0] ? $_[0] : '' );

    $_ = escape($_);
    s/([\x00-\x1f\x{7f}-\x{ffff}])/'\u'.chr2hex($1)/gse;

    return $_;
}

sub python_decoder {

    local $_ = ( defined $_[0] ? $_[0] : '' );

    s/((?:\A|\G|[^\\]))\\u([\da-fA-F]{4})/$1.hex2chr($2)/gse;

    return unescape($_);
}

#
# == common data and subroutines ==
#

my %ESCAPED = ( 
    "\\" => '\\', 
    "\r" => 'r', 
    "\n" => 'n', 
    "\t" => 't', 
    "\a" => 'a',
    "\b" => 'b',
    "\e" => 'e',
    "\f" => 'f',
    "\"" => '"',
    "\$" => '$',
    "\@" => '@',
);

my %UNESCAPED = ( reverse %ESCAPED );

sub escape ($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );
    s/([\a\b\e\f\r\n\t\"\\\$\@])/\\$ESCAPED{$1}/sg;
    return $_;
}

sub unescape ($) {
    local $_ = ( defined $_[0] ? $_[0] : '' );

    s/((?:\A|\G|[^\\]))\\([0-7]{1,3})/$1.oct2chr($2)/gse;

    s/((?:\A|\G|[^\\]))\\([^aAbBeEfFrRnNtT\\\"\$\@])/$1$2/g;

    s/((?:\A|\G|[^\\]))\\([aAbBeEfFrRnNtT\\\"\$\@])/$1.$UNESCAPED{lc($2)}/gse;

    return $_;
}



sub chr2hex {
    my($c) = @_;
    if ( ord($c) < 65536 ) {
        return sprintf("%04x", ord($c));
    }
    else {
        require Carp;
        Carp::croak (
            "'unicode-escape' codec can't encode character: ordinal " . ord($c)
        );
    }
}

sub hex2chr { 
    my($hex) = @_;
    if ( hex($hex) >= 0 and hex($hex) < 65536) {
        return chr(hex($hex));
    }
    else {
        require Carp;
        Carp::croak(
            "'unicode-escape' codec can't decode escape sequence: " 
            . "\\x$hex (ordinal " . hex($hex) . ")"
        );
    }
}

sub oct2chr {
    my($oct) = @_;
    if ( oct($oct) >= 0 and oct($oct) < 256 ) {
        return chr(oct($oct));
    }
    else {
        require Carp;
        Carp::croak (
            "'unicode-escape' codec can't decode escape sequence: " 
            . "\\$oct (ordinal " . oct($oct). ")"
        );
    }
}

$\ = "\n";


1;
__END__
=head1 NAME

Encode::Escape::Unicode - Perl extension for Encoding of Unicode Escape Sequnces

=head1 SYNOPSIS

  use Encode::Escape::Unicode;

  $escaped = "What is \\x{D384}? It's Perl!";
  $string = decode 'unicode-escape', $escaped;

  # Now, $string is equivalent "What is \x{D384}? It's Perl!"

  Encode::Escape::Unicode->demode('python');

  $python_unicode_escape = "And \\u041f\\u0435\\u0440\\u043b? It's Perl, too";
  $string = decode 'unicode-escape', $python_unicode_escape;

  # Now, $string eq "And \x{041F}\x{0435}\x{0440}\x{043B}? It's Perl, too"

If you have a text data file 'unicode-escape.txt'. It contains a line:

  Perl\tPathologically Eclectic Rubbish Lister\n

And you want to use it as if it were a normal double quote string in source 
code. Try this:

  open(FILE, 'unicode-escape.txt');
  while(<FILE>) {
    chomp;
    print decode 'unicode-escape', $_;
  }

=head1 DESCRIPTION

L<Encode::Escape::Unicode> module implements encodings of escape sequences.

Simply saying, it converts (decodes) escape sequences into Perl internal string 
(\x{0000} -- \x{ffff}) and vice versa.

See L<Encode::Escape> for more detailed description.

=head1 AUTHOR

you, E<lt>you at cpan dot orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by you 

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
# vi: set ts=4 sts=4 sw=4 et


lib/Encode/Escape.pm
# Encodings based on Escape Sequence

# $Id: Escape.pm,v 1.10 2007-12-05 14:17:20+09 you Exp $

package Encode::Escape;

use 5.008008;

use strict;
use warnings;

our $VERSION = do { q$Revision: 1.10 $ =~ /\d+\.(\d+)/; sprintf "%.2f", $1 / 100  };


sub import {
    if ( defined $_[1] and $_[1] eq ':modes') {

        require Exporter;

        our @ISA = qw(Exporter);
        our @EXPORT_OK = qw(enmode demode);

        __PACKAGE__->export_to_level(1, $_[0], 'enmode', 'demode');

        splice @_, 1, 1;
    }

    require Encode;
    Encode->export_to_level(1, @_);
}

use Encode::Escape::ASCII;
use Encode::Escape::Unicode;


sub enmode ($@) {
    my $enc = shift;
    my $obj = Encode::find_encoding($enc);

    unless (defined $obj) {

        require Carp;
        Carp::croak("Unknown encoding '$enc'");
    }

    $obj->enmode(@_);
}

sub demode ($@) {
    my $enc = shift;
    my $obj = Encode::find_encoding($enc);

    unless (defined $obj) {

        require Carp;
        Carp::croak("Unknown encoding '$enc'");
    }

    $obj->demode(@_);
}


1;
__END__

=head1 NAME

Encode::Escape - Perl extension for Encodings of various escape sequences

=head1 SYNOPSIS

  use Encode::Escape;

  $escaped_ascii = "Perl\\tPathologically Eclectic Rubbish Lister\\n";
  $ascii = decode 'ascii-escape', $escaped_ascii;
  
  # Now, $ascii is equivalent to 
  # double quote string "Perl\tPathologically Eclectic Rubbish Lister\n"

  $escaped_unicode = "Perl \\x{041F}\\x{0435}\\x{0440}\\x{043B} \\x{D384}" 
  $string = decode 'unicode-escape', $escaped_unicode;

  # Now, $string is equvialent to 
  # double quote string "Perl \x{041F}\x{0435}\x{0440}\x{043B} \x{D384}" 

It may looks non-sense. Here's another case.

If you have a text data file 'ascii-escape.txt'. It contains a line:

  Perl\tPathologically Eclectic Rubbish Lister\n

And you want to use it as if it were a normal double quote string in source 
code. Try this:

  open(FILE, 'ascii-escape.txt');
  while(<FILE>) {
    chomp;
    print decode 'ascii-escape', $_;
  }


=head1 DESCRIPTION

L<Encode::Escape> module is a wrapper class for encodings of escape sequences.

It is NOT for an escape-based encoding (eg. ISO-2022-JP). It is for 
encoding/decoding escape sequences, generally used in source codes.

Many programming languages, markup languages, and typesetting languages 
provide methods for encoding special (or functional) characters and non-keyboard 
symbols in the form of escape sequence. That's what I concern.

For example, L<Encode::Escape::Unicode> converts (encodes) Unicode characters
into escape sequences representing it: C<\x{AC00}> to the first
Hangul syllable Ka, C<\x{042F}> to Cyrillic capital letter Ya, etc.

Yes, you're right. There already exist many modules.
See L<String::Escape>, 
L<Unicode::Escape>, 
L<TeX::Encode>, 
L<HTML::Mason::Escape>, L<Template::Plugin::XML::Escape>, L<URI::Escape>, etc.

But for some reason I need to do it in a different way. There is more than 
one way to do it! After that, I asked myself if this module is useful. May 
be not except for me. At this time, Zhuangzi reminds me that L</"the useless 
has its use">.

=head2 Escape Sequences

=head3 Character Escape Codes

ASCII defines 128 characters: 33 non-printing control characters 
(0x00 -- 0x1f, 0x7f) and 95 printable characters (0x20 -- 0x7e).

Character Escape Codes in C programming language provide a method 
to express control characters, using only printable ones. 
These are accepted by Perl and many other languages.

 CEC    HEX     Description
 ---    ----    --------------
 \0     00      Null character
 \a     07      Bell
 \b     08      Backspace
 \t     09      Horizontal Tab
 \n     0a      Line feed
 \v     0b      Vertical Tab
 \f     0c      Form feed
 \r     0d      Carriage return

Programming languages provide escape sequences for printable characters,
which have significant meaning in that language. Otherwise, it would be
harder to print them literally.

 ESC    HEX     Description
 ---    ----    ---------------
 \"     22      double quote 
 \\     52      backslash

Refer to ASCII, Escape character, Escape sequence at <http://en.wikipedia.org>, for more
details.

=head3 Perl Escape Sequences

Perl use backslash as an escape character. 

These work in normal strings and regular expressions except \b.

 ESC       Description
 ---       --------------------------
 \a        Alarm (beep)
 \b        Backspace
 \e        Escape
 \f        Formfeed
 \n        Newline
 \r        Carriage return
 \t        Tab
 \037      Any octal ASCII value
 \x7f      Any hexadecimal ASCII value
 \x{263a}  A wide hexadecimal value
 \cx       Control-x
 \N{name}  name is a name for the Unicode character (use charnames)
 
The following escape sequences are available in constructs that interpolate.

 \l        Lowercase next character
 \u        Titlecase next character
 \L        Lowercase until \E
 \U        Uppercase until \E
 \E        End case modification
 
In regular expresssions:

 \b        An assetion, not backspace, except in a character class
 \Q        Disable pattern metacharacters until \E

Unlike C and other languages, Perl has no \v escape sequence for the vertical tab 
(VT - ASCII 11).

For constructs that do interpolate, variable begining with "$" or "@" are interpolated.

 \$        Dollar Sign
 \@        Ampersand
 \"        Print double quotes
 \         Escape next character if know otherwise print

See L<perlerref>, L<perlop>

=head3 Python Escape Sequences

 \newline   Ignored
 \\         Backslash (\)
 \'         Single quote (')
 \"         Double quote (")
 \a         ASCII Bell (BEL)
 \b         ASCII Backspace (BA)
 \f         ASCII Formfeed (FF)
 \n         ASCII Linefeed (LF)
 \N{name}   Character named 'name' in the Unicode database
 \r         ASCII Carriage Return (CR)
 \t         ASCII Horizontal Tab (TAB)
 \uxxxx     Character with 16-bit hex value xxxx (Unicode only)
 \Uxxxxxxxx Character with 32-bit hex value xxxxxxxx (Unicode only)
 \v         ASCII Vertical Tab (VT)
 \ooo       Character with octal value ooo
 \xhh       Character with hex value hh

See L<http://docs.python.org/ref/strings.html>

Unicode escape sequences in the form of C<\u>I<xxxx> are used in Java, Python,
C#, JavaScript. L<Unicode::Escape> module implements it.

=head3 LaTeX Escapes

L<TeX::Encode> implements encodings of LaTeX escapes. It converts (encodes)
utf8 string to LaTeX escapes.


=head1 SEE ALSO

=head1 The useless has its use 

Hui Tzu said to Chuang Tzu, "Your words are useless!"

Chuang Tzu said, "A man has to understand the useless before you can talk to 
him about the useful. The earth is certainly vast and broad, though a man 
uses no more of it than the area he puts his feet on. If, however, you were 
to dig away all the earth from around his feet until you reached the Yellow 
Springs, then would the man still be able to make use of it?"

"No, it would be useless," said Hui Tzu.

"It is obvious, then," said Chuang Tzu, "that the useless has its use."

=begin html

--- from <a href="http://www.terebess.hu/english/chuangtzu3.html"><em>External 
Things, Chuang Tzu translated by Burton Watson</em></a>

=end html

=head1 AUTHOR

You Hyun Jo, E<lt>you at cpan dot orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by You Hyun Jo

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
#set ts=4 sts=4 sw=4 et
